### ISSUE TITLE
Limit or reject slow clients

### ISSUE THREAD

**Issue by LexLuthr** [filecoin-project/boost#737]
With the connection limiting feature in boost, users can control how many simultaneous download can happen at once.
If a malicious or bad actor in the network makes multiple deals at once and all these deals are downloading at a very slow speed then Boost node will not be able to onboard more deals.
This slow down will results in good deals getting rejected or not being able to seal on time due to slow downloads of bad deals.

We need to make boost more robust against such bad deals and  allow some kind of limiter/rejected either based on current download speed of a deal or based on historical data for that client.

I think an outright rejection based on preset speed (set by user) for a limited time would be more robust but that can result in uptick of network chatter.

**Comment by scaseye** [comment]
estuary for example has at least 4 shuttles, the SPs connection speed to each shuttle will vary based on location.

is there a way to distinguish or limit based on the peer id in the transfer params?  estuary has one main peer id, but uses a sub peerid (the shuttle peerid) on the transfers.  

i would love to reject deals from the slow shuttles, and only except from the faster ones.

**Comment by dirkmc** [comment]
We're about to release an [HTTP Transfer Limiter](https://boost.filecoin.io/configuration/http-transfer-limit) that should help out somewhat. There's a detailed explanation there, but essentially it will favour faster connections over slower connections.

**Comment by jcace** [comment]
I'm having issues with this too. The HTTP Transfer Limiter seems like a good approach, do you know when/what version of Boost it will be released in?

**Comment by LexLuthr** [comment]
@jcace We will cutting a release candidate tomorrow. Boost v1.4.0-rc2

**Comment by jcace** [comment]
Just tested this out in Boost v1.4.0-rc2, works wonderfully. Nice work!

My only concern with it is I'm still running in to an issue where the queue is clogged up with a bunch of slow transfers. I have `LOTUS_DEALMAKING_HTTPTRANSFERMAXCONCURRENTDOWNLOADS` set to `10`, but now all 10 of those transfers are going slowly, will take several hours to complete, blocking faster transfers from getting picked up. 

I read that the algorithm will prefer deals with peers that don't currently have ongoing transfers, but I wonder if it would also be useful to implement another configuration parameter, i.e, `MaxConcurrentDownloadsPerPeer` which could be set to limit transfers within a single peer only, instead of a limit applied to all transfers. This way the download queue could not be saturated with slow transfers and could leave room for some other peers.

**Comment by dirkmc** [comment]
@jcace the algorithm tries to strike a balance between servicing deals in FIFO order and choosing fast peers:
we don't want to get bogged down by slow peers, but we also don't want to _always_ choose the fast peers because then older deals from slow peers won't get downloaded in time to meet the required start epoch (in the deal proposal) and the deal will fail.

I'd like to avoid a per-peer limit because I think it might end up causing other problems, as mentioned in the comment you linked to. However we may end up needing it in practice. Please do update us on how it goes for you.

**Comment by jcace** [comment]
Been running for the past 2 days - unfortunately we haven't noticed much of an improvement over the previous behavior. 

This concurrent download limit is good in theory, but unfortunately it doesn't solve the problem caused by the Estutary shuttle that's feeding us deals. The problem is, this shuttle has:

1. slow download speed to our node (between 20-40Mbps at any one time)
2. a seemingly unlimited amount of deals

The average deal size is 8.8 GiB. So, if we have 10 of those running at a time and the 30Mbps gets split between them, 8.8Gib @ 3Mbps = ~7 hours to complete the download. Some downloads are even larger ~28Gib, those ones will take 20+ hours to complete. 

So, our queue is often full with 10 of these "slow" deals downloading, and we may have to wait up to 7+ hours for one to complete and another slot to open up. In that timespan, other deals will show up that can be downloaded from a local server at 100x the speed, but they can't start as the queue is full. The end result is that our 10Gbps internet uplink is usually only utilized at 30Mbps 😢, and we can't ingest as many deals as possible due to the slow ones still blocking the fast ones.

I think the sheer difference in download speed between the shuttles is the source of this issue (local deals can download at 400Mbps+, completing in 4 minutes vs. 3Mbps/7 hours for the slow ones).

With that said, there are still benefits to this concurrent limit as now those slow downloads actually have a chance at completing, but I do think a concurrent limit **per peer** would ultimately resolve the issue we're currently seeing. Or, some other mechanism to de-prioritize that slow shuttle and always keep some slots open for faster deals.

**Comment by jacobheun** [comment]
@benjaminh83 is seeing similar issues as mentioned in https://github.com/filecoin-project/boost/pull/710#issuecomment-1236043586. I know Estuary is working on stability improvements to help mitigate this issue. I would like to ensure we're able to prevent this growing backlog of deals for any future clients though. 

I'd like to understand the acceptance flow a bit better here, because the resolution in my mind here is to start rejecting deals once a concurrency limit is exceeded. If we don't do this we'll end up with a queue that has no hope of draining in time to make the publish deadline. Effectively, I think we should be doing the equivalent of an HTTP 429, with the ability to allow certain, performant peers to have higher limits.

@jcace @benjaminh83, can you do per peer, concurrent **deal** limits with tools like CID Gravity today? Are their implications you want to avoid by outright rejecting deals once they get to a certain threshold (ex: 10 concurrent deals that have not finished transferring)?

A per peer limit doesn't seem like it would fully solve this problem, because we're still accepting deals that are ultimately going to fail.

**Comment by jcace** [comment]
Hey again guys. I did a bit of experimenting - forked boost v1.4.0-rc2 and added in per-peer concurrent limiting to the codebase through another environment variable. The performance improvement was immediately noticeable for our situation - fast downloads started immediately while slow ones continued along. End result was more verified deals entering our SP. 

<img width="489" alt="Capture" src="https://user-images.githubusercontent.com/12418265/188779220-231a9c82-1239-4dc9-a42d-6e58c9f0e5a0.PNG">

@jacobheun  Yes, I'm still seeing an issue with deals stacking up in a **transfer queued** state, and some of them expiring before the publish deadline. I agree with your thoughts about rejecting deals if there's no hope of reaching that deal in time. However I'd want to make sure there's a happy balance with keeping some ready to go once transfers complete. 
Naively, maybe we allow 2x the concurrent transfer limit to queue up before rejecting additional deals? Ideally there would be some notion of transfer speed / de-queueing rate that could be used to more accurately size the queue. 

@dirkmc I recognize that per-peer limit may not be the solution for everyone, but I'm more than happy to submit a PR if you'd like 🙂

https://github.com/jcace/boost/tree/feat/http-per-peer-xferlimit

**Comment by dirkmc** [comment]
@jcace oh nice, thanks for putting together this experiment 🙏  We always welcome contributions.
It does sound like the per-peer limit is going to be useful. I just want to make sure we think about it comprehensively so we can make sure we cover the different use cases in a way that's applicable to all SPs if possible.

I have been thinking about this issue the last couple of days. Conceptually we have a situation something like the following:

<img width="723" alt="Screen Shot 2022-09-07 at 4 31 22 PM" src="https://user-images.githubusercontent.com/169124/188905029-cd74ab37-9e98-4c86-8ecc-6fdcda69059a.png">

- The sealing pipeline has a maximum capacity, eg *128GB / hour*
- The space to store incoming downloads should be larger than the sealing pipeline capacity so as to provide a buffer for bursty traffic from the internet, eg 10 hours worth: *1280GB*
- In this scenario there are two hosts, one with a 3mbps connection and the other with an 8mbps connection.
- Some data is transferring through each pipe, and some data has been accepted and queued for transfer from each host (but is not yet transferring).
- A new storage deal proposal has arrived whose data is on host A.

In broad terms I'd suggest we:
1. Limit the size of the accepted deals queue per-host
  This will solve the problem of a slow host sending the SP one thousand deals, which reserves all the space in the incoming downloads buffer and prevents the SP from being able to accept deals from a host with a faster connection.
  For example the accepted deals queue limit could be set to 20% of the size of the incoming downloads buffer, per peer. Once the limit is reached, the SP rejects further deals from that host, but still has room to accept deals from another (faster) host.
2. Order deal proposals by Deal Proposal Start Epoch (instead of deal proposal creation time)
  In practice it's most important to prioritize downloading deals which must be started earliest.
3. Come up with a smarter algorithm for choosing which deals to download
  The current algorithm is designed to address the problem of stalled peers, but in practice it seems the problem is slow peers.
  The solution could be a per-peer limit as suggested by @jcace, or another solution. This one is most complex so I suggest we start a separate issue to discuss.

If we immediately implement the accepted deals queue per-host size limit I think it will address the most pressing issue. Then we can spend some more time working out the details of the smarter transfer algorithm.

**Comment by jcace** [comment]
Nice breakdown, thanks @dirkmc ! I agree with all the points above and your suggestions #1 and #2 make sense.

> The space to store incoming downloads should be larger than the sealing pipeline capacity so as to provide a buffer for bursty traffic from the internet, eg 10 hours worth: 1280GB 

In theory isn't this handled by the `LOTUS_DEALMAKING_MAXSTAGINGDEALSBYTES` config? It seems that the code already takes into account the amount of storage space dedicated for the staging area and reserves it before starting the download. 

Maybe it's a different issue, but this staging byte size doesn't perform that function now, as a lot of old deals that have failed/expired are still clogging up the space. if I want to free up that space, I need to go through this and click all the "X" hundreds of times across many pages of deals. Perhaps the deal reject functionality, combined with a cleanup/age-out of old deals would accomplish this and still make it easy for SP's to size their staging area appropriately based on their sealing speed./

<img width="1065" alt="Screen Shot 2022-09-07 at 1 48 33 PM" src="https://user-images.githubusercontent.com/12418265/188975701-dc4aab3a-3bb3-4063-935d-978fe822d414.png">


ex) I have my staging size set to 30TiB but most of that is from these old deals that are still hanging around from several weeks ago
<img width="1063" alt="Screen Shot 2022-09-07 at 1 53 25 PM" src="https://user-images.githubusercontent.com/12418265/188976563-50d99032-c8be-4f74-b2bd-da18d718ba18.png">

**Comment by dirkmc** [comment]
@jcace yes you're right, we already have a config option for the size of the incoming downloads staging area 👍 
I was just noting that SPs should think about the size of this space relative to the size of their sealing pipeline.

With regards to having to manually clear out the deals that have failed to download, there are three options for when something goes wrong with a deal:
1. "Pause" the deal in it's current state, and require the user to manually restart or cancel the deal (even if the node is restarted)
2. "Pause" the deal in it's current state, allow the user to intervene manually, and restart the deal when the node is restarted
3. Immediately fail the deal

At the moment when a transfer fails because it has timed out after many attempts, we pause it and require the user to manually restart or cancel the deal. It seems like this is annoying for SPs to have to do in practice, so instead I will change the code to immediately fail the deal when the transfer fails.

Thanks for giving quick feedback, and raising these issues you're having, it really helps us to improve the product for everyone 🙏

**Comment by dirkmc** [comment]
@jcace by the way if you want to filter for deals by error message, you can use the search box at the top right of the Deals page to search for the error message (eg search for "data-transfer failed"

**Comment by dirkmc** [comment]
Opened PR [Configurable download limit per host](https://github.com/filecoin-project/boost/pull/780#top) to allow SPs to limit the size of the accepted deals queue per-host

**Comment by dirkmc** [comment]
@jcace I've merged [Configurable download limit per host](https://github.com/filecoin-project/boost/pull/780) into `main`. Would you be able to try it out and let us know if it helps?

**Comment by jcace** [comment]
Hi @dirkmc  , just some quick questions about the behaviour before I switch over. 

1. Is there any mechanism to auto age-out / remove old deals yet? I spammed the "X" button on a bunch but i still have 1500+ of them , several weeks old. Will those contribute to the download limit and fill it up? Should I go through and completely remove all of them first?

2. The max. concurrent download limit still applies, in addition to the new staging deals storage space limit? 

3. Screenshot below, showing what Boost is doing currently. When it starts back up, will it pick up those 10 active ones again and then cut down the 128 queued deals to a smaller amount based on the %-per-host I configure? 
<img width="657" alt="Screen Shot 2022-09-12 at 4 44 14 PM" src="https://user-images.githubusercontent.com/12418265/189777069-8ba1e19d-1792-4d30-b17b-5c99d30f70c9.png">
s? 

Thanks again for your work on this

**Comment by dirkmc** [comment]
@jcace the answers to 2 and 3 are yes, that's the expected behaviour. Let us know if it works for you.

The answer to 1 is that you can use the [graphql API](https://boost.filecoin.io/graphql-api) to search for deals and cancel them from the command line.
To filter deals using a search query against the deal message you can use a query like:
```
curl -X POST \
-H "Content-Type: application/json" \
-d '{"query":"query { deals(limit: 10, query: \"search term goes here\") { deals { ID CreatedAt Message } } }"}' \
http://localhost:8080/graphql/query | jq
```
For example the search term could be "data-transfer failed".

To cancel a deal from the command line use:
```
curl -X POST \
-H "Content-Type: application/json" \
-d '{"query":"mutation { dealCancel(id: \"ab12345c-5678-90de-12f3-45a6b78cd9ef\") }"}' \
http://localhost:8080/graphql/query | jq
```
where "ab12345c-5678-90de-12f3-45a6b78cd9ef" is the deal ID.

**Comment by jcace** [comment]
Working on this now...

FYI the mutation to terminate the deal is `dealFailPaused(id: ..)` not `dealCancel()`. The `dealCancel` mutation appears to just stop an in-progress download but does not remove/terminate the deal from the queue

**Comment by scaseye** [comment]
would be really great to be able to control the release of published deals to the sealing system based on what is currently running on the sealing system.  I do this on offline deals via scripts.  for me and i think most SPs PC1 is the bottleneck. 

my script does the following prior to importing the next deal.  Would be awesome if boost did the same thing _prior to sending a deal to be published_.  Once the boost stage area is full, then start rejecting deals.

MAXSEALJOBS=50
MAXPUBLISH=6

while DEALS in Boost Staging

check_pendPublish (if < MAXPUBLISH continue)
check_mpool (if MPOOL empty continue)
SEALJOBS = $(lotus-miner sealing jobs | grep -c "PC1\|RU\|GET\|FRU")
check_jobs (  IF SEALJOBS < MAXSEALJOBS continue)
PUBLISH 


```
check_mpool(){
	mps=$(lotus mpool pending --local | wc -l)
	echo checking mpool for msgs, found "$mps"
	while [ "$mps" -gt 0 ]
	do
		echo Found msgs in mpool sleep 15 seconds...
		sleep 15
		mps=$(lotus mpool pending --local | wc -l)
	done
}
check_jobs(){
	if [ $runCount -gt $MAXRUN ]
	then 
		echo Max Run of successful sectors done!
		exit 
	fi
	PC1s=$(lotus-miner sealing jobs | grep -c "PC1\|RU\|GET\|FRU")
	#PC1s=$(lotus-miner sealing jobs | grep -c "RU\|GET\|FRU")
	echo "Checking RU jobs! Found $PC1s, Max set to $MAXPC1 ."
	while [ "$PC1s" -gt $MAXPC1 ]
	do
		echo "to many Seal Jobs to contintue sleeping for $SLEEPJOBCHECK seconds..."
		sleep $SLEEPJOBCHECK
		PC1s=$(lotus-miner sealing jobs | grep -c "PC1\|RU\|GET\|AP\|FRU")
		echo "Checking SEAL jobs! Found $PC1s, Max set to $MAXPC1."
	done
}
check_pendPublish(){
	publishQueue=$(lotus-miner storage-deals list | grep -c StorageDealStaged)
	echo "Checking StorageDealStage! Found $publishQueue, Max set to $MAXPB ."
	while [ "$publishQueue" -gt $MAXPB ]
	do
		echo "to many deals in StorageDealStage Queue to contintue sleeping for $SLEEPJOBCHECK seconds..."
		sleep $SLEEPJOBCHECK
		publishQueue=$(lotus-miner storage-deals pending-publish |grep -c bafyrei)
		echo "Checking StorageDealStage! Found $publishQueue, Max set to $MAXPB ."
	done
}

```

**Comment by scaseye** [comment]
i think the [http-per-peer-xferlimit] would be really helpful.   CIDGravity can limit Estuary but not its sub shuttles.  Or at least not right now.  

also would like to reject by host too.  using estuary's example, I would like to not allow any deals from the shuttles that are across the world and will never have good speeds.

that plus being able to allow more connections from known fast shuttles, and less from slow ones would really improve throughput.

**Comment by jcace** [comment]
@dirkmc I've been testing for 3 days now and things are excellent. No more massive list of 1000's of dead transfers, they clear out automatically. The % and staging area system takes a bit of trial and error to get the numbers correct - but maybe that's because I'm working backwards from a # of concurrent transfers per-host (due to Estutary shuttle slowness issue). 

Only thing I'm a bit confused about is the difference between the two config items, `LOTUS_LOTUSDEALMAKING_MAXSTAGINGDEALSBYTES` and `LOTUS_DEALMAKING_MAXSTAGINGDEALSBYTES`. I have the new one (LotusDealmaking) set to 3TiB and the older one (Dealmaking) set to 30TiB. I see the 30TiB storage space in the Boost sidebar, but when will this ever get used? Assuming the smaller, new config item will override it / fill up first before getting close to the full 30TiB?

<img width="231" alt="Screen Shot 2022-09-18 at 9 59 16 AM" src="https://user-images.githubusercontent.com/12418265/190919191-09b77446-e06b-44b0-820c-d79e899dc353.png">

**Comment by dirkmc** [comment]
@scaseye I'm going to close this issue because we've resolved most of the problems, and I think it would be clearer to start a new issue with the remaining problem that you're seeing.
Could you please open a new issue outlining the exact problem that you're seeing. Then in the comments on the issue we can discuss different solutions to the problem.
